### DISCUSSION TITLE
Keda Operator and Metrics API Server unable to start in AKS

### DISCUSSION THREAD

**Discussion by MarekLani** [kedacore/keda#3137] [Q&A / Need Help] [answered]
### Report

When installing kEDA thru helm chart onto AKS cluster (behind firewall) we are getting following errors from keda operator and metrics api server:

```
1.6545953013458195e+09 ERROR Failed to get API Group-Resources {"error": "Get \"https://10.0.0.1:443/api?timeout=32s\": EOF"}
sigs.k8s.io/controller-runtime/pkg/cluster.New
/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/cluster/cluster.go:160
sigs.k8s.io/controller-runtime/pkg/manager.New
/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/manager/manager.go:313
main.main
/workspace/main.go:87
runtime.main
/usr/local/go/src/runtime/proc.go:255
1.6545953013459463e+09 ERROR setup unable to start manager {"error": "Get \"https://10.0.0.1:443/api?timeout=32s\": EOF"}
main.main
/workspace/main.go:97
runtime.main
/usr/local/go/src/runtime/proc.go:255
```

```
I0607 09:53:05.297924 1 main.go:147] keda_metrics_adapter "msg"="KEDA Version: 2.7.1"
I0607 09:53:05.297979 1 main.go:148] keda_metrics_adapter "msg"="KEDA Commit: "
I0607 09:53:05.297996 1 main.go:149] keda_metrics_adapter "msg"="Go Version: go1.17.9"
I0607 09:53:05.298006 1 main.go:150] keda_metrics_adapter "msg"="Go OS/Arch: linux/amd64"
E0607 09:53:15.344324 1 logr.go:279] keda_metrics_adapter "msg"="Failed to get API Group-Resources" "error"="Get \"https://10.0.0.1:443/api?timeout=32s\": EOF"
E0607 09:53:15.344360 1 main.go:104] keda_metrics_adapter "msg"="failed to setup manager" "error"="Get \"https://10.0.0.1:443/api?timeout=32s\": EOF"
E0607 09:53:15.344378 1 main.go:209] keda_metrics_adapter "msg"="making provider" "error"="Get \"https://10.0.0.1:443/api?timeout=32s\": EOF"
E0607 09:53:15.344399 1 main.go:168] keda_metrics_adapter "msg"="unable to run external metrics adapter" "error"="Get \"https://10.0.0.1:443/api?timeout=32s\": EOF"
```

### Expected Behavior

After running helm install KEDA initializes properly.

### Actual Behavior

Keda Operator and Keda Metrics API Server pods are getting into crash loop backoff state.

### Steps to Reproduce the Problem

1. AKS Cluster behind Firewall
2. Deploy KEDA thru Helm Chart

### Logs from KEDA operator

As stated above

```
1.6545953013458195e+09 ERROR Failed to get API Group-Resources {"error": "Get \"https://10.0.0.1:443/api?timeout=32s\": EOF"}
sigs.k8s.io/controller-runtime/pkg/cluster.New
/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/cluster/cluster.go:160
sigs.k8s.io/controller-runtime/pkg/manager.New
/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/manager/manager.go:313
main.main
/workspace/main.go:87
runtime.main
/usr/local/go/src/runtime/proc.go:255
1.6545953013459463e+09 ERROR setup unable to start manager {"error": "Get \"https://10.0.0.1:443/api?timeout=32s\": EOF"}
main.main
/workspace/main.go:97
runtime.main
/usr/local/go/src/runtime/proc.go:255
```

### KEDA Version

2.7.1

### Kubernetes Version

1.22

### Platform

Microsoft Azure

### Scaler Details

N/A

**Comment by tomkerkhove** [comment, marked as answer]
Did you verify that the ports used are enabled? Most probably that's the issue

**Reply by MarekLani** [reply]
@tomkerkhove  thank you for your quick response. However please could you elaborate bit more. Is there some communication needed to happen outside of the cluster, when KEDA is being instantiated? I understood https://10.0.0.1:443/ target in within cluster

**Reply by tomkerkhove** [reply]
You mentioned that it's behind a firewall so the firewall might block the ports.

This is what is required: https://keda.sh/docs/2.7/operate/cluster/#firewall

**Reply by MarekLani** [reply]
Understand now, needed to explore the actual routing and indeed despite it is kube api server it seems firewall stands between the pods and the endpoint. Thanks also for the article.

**Reply by tomkerkhove** [reply]
Happy to help! Did this fix your issue?

**Reply by MarekLani** [reply]
I will report back, unfortunately our team has no management rights over FW and it needs to be solved by raising ticket.  Thank you again!

**Reply by MarekLani** [reply]
We have configured network rule on our FW as suggested here: https://docs.microsoft.com/en-us/azure/aks/limit-egress-traffic#azure-global-required-network-rules and KEDA now initializes successfully. Thank you for support once again.

**Reply by tomkerkhove** [reply]
Good to hear! Happy scaling.
